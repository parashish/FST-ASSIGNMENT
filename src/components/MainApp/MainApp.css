body {
  background-color: #e5eaef;
}
.whole-loader {
  left: 50%;
  margin-left: -10em;
  position: fixed;
  top: 50%;
}
.cards {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}
.main-wrapper {
  display: block;
  margin: 20px;
}

.filter-section {
  width: 100%;
}
.main-section {
  width: 100%;
}
@media (min-width: 768px) and (max-width: 1024px) {
  .main-wrapper {
    display: flex;
  }
  .filter-section {
    width: 25%;
    margin-right: 20px;
  }
  .main-section {
    width: 75%;
  }
}

@media only screen and (min-width: 1024px) {
  .main-wrapper {
    display: flex;
  }
  .filter-section {
    width: 25%;
    margin-right: 20px;
  }
  .main-section {
    width: 75%;
  }
}
