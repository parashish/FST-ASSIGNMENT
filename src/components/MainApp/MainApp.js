import React from 'react';
import './MainApp.css';
import Header from '../Header/Header';
import Filters from '../Filters/Filters';
import CharacterList from '../CharacterList/CharacterList';
import axios from 'axios';
import Loader from '../../assets/icons/loader-animation.gif';

class MainApp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      characters: [],
      sortById: 'asc',
      width: 0,
      height: 0,
      showFiltersOnMobile: false,
      showOnlyLoader: true,
      filterOptions: {
        species: [],
        gender: [],
        origin: [],
        names: [],
      },
      selectedSpecies: [],
      selectedGender: [],
      selectedOrigin: [],
      selectedNames: [],
      nameFilterOptions: '',
    };
    window.removeEventListener('resize', this.updateWindowDimensions);
  }

  componentDidMount() {
    this.getCharacterDetails();
    this.updateWindowDimensions();
    window.addEventListener('resize', this.updateWindowDimensions);
  }

  render() {
    return this.state.showOnlyLoader ? (
      <img className="whole-loader" src={Loader} alt="Loading..." />
    ) : (
      <div className="main-wrapper">
        <div className="filter-section">
          <Filters
            selected={{
              species: this.state.selectedSpecies,
              gender: this.state.selectedGender,
              origin: this.state.selectedOrigin,
            }}
            onChangeCheckboxObject={this.onChangeCheckboxObject}
            filterOptions={this.state.filterOptions}
            onClickShowFilters={this.onClickShowFilters}
            screenSize={{ height: this.state.height, width: this.state.width }}
            showFiltersOnMobile={this.state.showFiltersOnMobile}
            onChangeObject={this.onChangeObject}
          />
        </div>
        <div className="main-section">
          {(this.state.width >= 768 || this.state.showFiltersOnMobile) && (
            <Header
              onRemoveItem={this.onRemoveItem}
              onChangeCheckboxObject={this.onChangeCheckboxObject}
              selected={{
                species: this.state.selectedSpecies,
                gender: this.state.selectedGender,
                origin: this.state.selectedOrigin,
                names: this.state.selectedNames,
              }}
              filterOptions={this.state.filterOptions}
              sortById={this.state.sortById}
              onChangeObject={this.onChangeObject}
              onSearch={this.onSearch}
              nameFilterOptions={this.state.nameFilterOptions}
            />
          )}
          <div className="cards">
            {this.state.characters
              .filter((character) => {
                return (
                  (this.state.selectedSpecies.length === 0 ||
                    this.state.selectedSpecies.indexOf(character.species) !==
                      -1) &&
                  (this.state.selectedGender.length === 0 ||
                    this.state.selectedGender.indexOf(character.gender) !==
                      -1) &&
                  (this.state.selectedOrigin.length === 0 ||
                    this.state.selectedOrigin.indexOf(character.origin.name) !==
                      -1) &&
                  (this.state.selectedNames.length === 0 ||
                    this.state.selectedNames.indexOf(
                      character.name.split(' ')[0],
                    ) !== -1 ||
                    this.state.selectedNames.indexOf(
                      character.name.split(' ')[1],
                    ) !== -1 ||
                    this.state.selectedNames.indexOf(character.name) !== -1)
                );
              })
              .sort((a, b) => {
                if (this.state.sortById === 'asc' && a.id > b.id) return 1;
                else return -1;
              })
              .map((character, idx) => {
                return <CharacterList key={idx} character={character} />;
              })}
          </div>
        </div>
      </div>
    );
  }

  /**
   * @description - API call to get character list and set result in component state
   */
  getCharacterDetails = () => {
    axios.get(`https://rickandmortyapi.com/api/character`).then((response) => {
      let newState = this.state;
      const results = response.data.results;
      newState.characters = response.data.results;
      newState.showOnlyLoader = false;
      newState.filterOptions.species = results
        .map((character) => {
          return character.species;
        })
        .filter(this.onlyUnique);
      newState.filterOptions.gender = results
        .map((character) => {
          return character.gender;
        })
        .filter(this.onlyUnique);
      newState.filterOptions.origin = results
        .map((character) => {
          return character.origin.name;
        })
        .filter(this.onlyUnique);
      newState.filterOptions.names = results
        .map((character) => {
          return character.name;
        })
        .filter(this.onlyUnique);
      this.setState(newState);
    });
  };

  /**
   * @param {HTMLElement}
   * @description - on change handler
   */
  onChangeObject = (e) => {
    let newState = this.state;
    newState[e.target.name] = e.target.value;
    this.setState(newState);
  };

  /**
   * @param {HTMLElement}
   * @description - on change handler
   */
  onChangeCheckboxObject = (e) => {
    let newState = this.state;
    if (this.state[e.target.name].indexOf(e.target.value) === -1) {
      newState[e.target.name].push(e.target.value);
    } else {
      newState[e.target.name].splice(
        this.state[e.target.name].indexOf(e.target.value),
        1,
      );
    }
    this.setState(newState);
  };

  /**
   * @param {HTMLElement}
   * @description - on change handler
   */
  updateWindowDimensions = () => {
    this.setState({ width: window.innerWidth, height: window.innerHeight });
  };

  /**
   * @description - show filter in mobile
   */
  onClickShowFilters = () => {
    this.setState({ showFiltersOnMobile: !this.state.showFiltersOnMobile });
  };

  onlyUnique = (value, index, self) => {
    return self.indexOf(value) === index;
  };

  /**
   * @param {String} - Filter Name
   * @param {String} - Type of filter
   * @description - Remove selected filter from the list
   */
  onRemoveItem = (name, type) => {
    let newState = this.state;
    switch (type) {
      case 'selectedSpecies':
        newState.selectedSpecies.splice(
          newState.selectedSpecies.indexOf(name),
          1,
        );
        break;
      case 'selectedGender':
        newState.selectedGender.splice(
          newState.selectedGender.indexOf(name),
          1,
        );
        break;
      case 'selectedOrigin':
        newState.selectedOrigin.splice(
          newState.selectedOrigin.indexOf(name),
          1,
        );
        break;
      case 'selectedNames':
        newState.selectedNames.splice(newState.selectedNames.indexOf(name), 1);
        break;
      default: {
      }
    }
    this.setState(newState);
  };
  /**
   * @param {HTMLElement} -
   * @description - Search functionality on the basis of name
   */
  onSearch = (e) => {
    let newState = this.state;
    if (
      this.state[e.target.name].indexOf(e.target.previousSibling.value) ===
        -1 &&
      e.target.previousSibling.value.length > 3
    ) {
      newState[e.target.name].push(e.target.previousSibling.value);
    }
    this.setState(newState);
  };
}

export default MainApp;
