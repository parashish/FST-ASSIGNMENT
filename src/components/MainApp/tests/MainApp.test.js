import React from 'react';
import { shallow } from 'enzyme';
import MainApp from '../MainApp';
import axios from 'axios';

jest.mock('axios', () => {
  const response = {
    data: {
      results: [
        {
          id: 1,
          name: 'Rick Sanchez',
          status: 'Alive',
          species: 'Human',
          type: '',
          gender: 'Male',
          origin: {
            name: 'Earth (C-137)',
            url: 'https://rickandmortyapi.com/api/location/1',
          },
          location: {
            name: 'Earth (Replacement Dimension)',
            url: 'https://rickandmortyapi.com/api/location/20',
          },
          image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
          episode: [
            'https://rickandmortyapi.com/api/episode/1',
            'https://rickandmortyapi.com/api/episode/2',
          ],
          url: 'https://rickandmortyapi.com/api/character/1',
          created: '2017-11-04T18:48:46.250Z',
        },
        {
          id: 2,
          name: 'Morty Smith',
          status: 'Alive',
          species: 'Human',
          type: '',
          gender: 'Male',
          origin: {
            name: 'Earth (C-137)',
            url: 'https://rickandmortyapi.com/api/location/1',
          },
          location: {
            name: 'Earth (Replacement Dimension)',
            url: 'https://rickandmortyapi.com/api/location/20',
          },
          image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
          episode: [
            'https://rickandmortyapi.com/api/episode/1',
            'https://rickandmortyapi.com/api/episode/2',
          ],
          url: 'https://rickandmortyapi.com/api/character/2',
          created: '2017-11-04T18:50:21.651Z',
        },
      ],
    },
  };
  return {
    get: jest.fn(() => Promise.resolve(response)),
  };
});
describe('MainApp component', () => {
  it('should render correctly with props` and match snapshot', () => {
    const component = shallow(<MainApp />);
    expect(component).toMatchSnapshot();
  });
  it('should render correctly with props` and match snapshot', () => {
    const component = shallow(<MainApp />);
    const instance = component.instance();
    jest.spyOn(instance, 'getCharacterDetails');
    instance.componentDidMount();
    expect(instance.getCharacterDetails).toHaveBeenCalledTimes(1);
  });
});
