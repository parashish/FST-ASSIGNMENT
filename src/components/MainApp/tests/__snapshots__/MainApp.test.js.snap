// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainApp component should render correctly with props\` and match snapshot 1`] = `
<img
  alt="Loading..."
  className="whole-loader"
  src={Object {}}
/>
`;

exports[`MainApp component should render correctly with props\` and match snapshot 2`] = `
<img
  alt="Loading..."
  className="whole-loader"
  src={Object {}}
/>
`;
